pkg/genericapiserver/destinations.go

```diff
@@ -0,0 +1,189 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package genericapiserver
+
+import (
+	"fmt"
+
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/runtime/serializer/versioning"
+	"k8s.io/kubernetes/pkg/storage"
+	etcdstorage "k8s.io/kubernetes/pkg/storage/etcd"
+	"k8s.io/kubernetes/pkg/util/sets"
+
+	"github.com/golang/glog"
+)
+
+// StorageDestinations is takes a GroupResource and returns back its storage interface.  This result includes:
+// 1. Merged etcd config, including: auth, server locations, prefixes
+// 2. Resource encodings for storage: group,version,kind to store as
+// 3. Cohabitating default: some resources like hpa are exposed through multiple APIs.  They must agree on 1 and 2
+type StorageDestinations struct {
+	// DefaultEtcdConfig describes how to connect to etcd in general.  It's authentication information will be used for
+	// every storage.Interface returned.
+	DefaultEtcdConfig etcdstorage.EtcdConfig
+
+	// EtcdLocationOverrides contains the list of "special" locations that are used for particular GroupResources
+	// these are merged on top of the default DefaultEtcdConfig when requesting the storage.Interface for a given GroupResource
+	EtcdLocationOverrides map[unversioned.GroupResource][]string
+
+	// EtcdPrefixOverrides contains the list of "special" prefixes for a GroupResource.  Resource=* means for the entire group
+	EtcdPrefixOverrides map[unversioned.GroupResource]string
+
+	// DefaultSerializer is used to create encoders and decoders for the storage.Interface.
+	DefaultSerializer runtime.NegotiatedSerializer
+
+	// SerializerOverrides contains the list of "special" serializers for a GroupResource.  Resource=* means for the entire group
+	SerializerOverrides map[unversioned.GroupResource]runtime.NegotiatedSerializer
+
+	// ResourceEncodingConfig describes how to encode a particular GroupVersionResource
+	ResourceEncodingConfig ResourceEncodingConfig
+
+	// CohabitatingResources keeps track of which resources must be stored together.  This happens when we have multiple ways
+	// of exposing one set of concepts.  autoscaling.HPA and extensions.HPA as a for instance
+	// the order of the slice matters!  its the priority order of lookup for finding a storage location
+	CohabitatingResources map[unversioned.GroupResource][]unversioned.GroupResource
+
+	// ResourceEnablement indicate whether the *storage* is enabled, NOT the API
```

Comment:
It's a little difficult, since the config grew organically and is a mix of default on and default off.

In order to support default on, you have to have default encodings, locations, prefixes for new resources.  Sometimes however there are explicit overrides and offs (this is currently inconsistently as a mix of flags, files, and code).

This code resolves the discrepancies by splitting the responsibilities.  Encoding, location, and prefix are all default on.  Which one to use in cohabitation cases is done by a separate enablement that (currently) is some default on and some default off.

We want the interface because in openshift we'll build a prescriptive set of enablement rules based on an explicit config plus defaulting rules.  This interface will let us do that and make it easier to rework specific parts of the config in kube too.   The current code is hard to work with.